File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.java
class Solution {
    public int[][] solution(int n) {
        int[][] answer = new int[n][n];
        int num = 1;
        int x = 0;
        int y = 0;
        int min = 0;
        int max = n - 1;
        boolean a = false;
        boolean b = false;
        
        while(num <= n*n){
            answer[y][x] = num;
            
            if(x == max && y == min){
                a = true;
            }else if(x == max && y == max){
                b = true;
            }else if(x == min && y == max){
                a = false;
                min++;
            }else if(x == min - 1 && y == min){
                b = false;
                max--;
            }
            
            if(!a && !b){
                x++;
                num++;
            }else if(a && !b){
                y++;
                num++;
            }else if(a && b){
                x--;
                num++;
            }else if(!a && b){
                y--;
                num++;
            }
            
        }
        
        return answer;
    }
}

File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

int** solution(int n) {
    int** answer = (int**)malloc(n * sizeof(int*));
    int cnt = 1;
    int dir = 0;
    int x = 0, y = 0;
    int nx = 0, ny = 0;
    int xx[4] = {1, 0, -1, 0};  // 우, 하, 좌, 상
    int yy[4] = {0, 1, 0, -1};
    
    for(int i = 0; i < n; i ++){
        answer[i] = calloc(n, sizeof(int));
    }
    
    while(cnt <= n * n){
        answer[y][x] = cnt++;
        
        // xx 또는 yy 배열로 다음 좌표계산
        nx = x + xx[dir];
        ny = y + yy[dir];
        
        // x 또는 y가 0보다 작아질때, n-1보다 클때, 이미 삽입한 값이 있을때 방향 변경
        if(nx < 0 || nx >= n || ny < 0 || ny >= n || answer[ny][nx] != 0){
            dir = (dir + 1) % 4;    // 처음 0 - 1 - 2 - 3 - 0 ...
            nx = x + xx[dir];
            ny = y + yy[dir];
        }
        
        x = nx;
        y = ny;
    }
    
    
    return answer;
}